### PR TITLE
Explicit rest_communication retries

### LINKED ISSUE #105: SSL errors in rest_communication
#91 was fixed, but we are still seeing this occur at random. We should try to reproduce the error and possibly stop caching the connection object

### PR DESCRIPTION
**Pull request by mwhamgenomics**
Uses our own retry mechanism - closes #105

### CODE REVIEW

**Comment by mwhamgenomics** [comment]
Previously with the urllib3 Retry and a backoff factor of 0.2, the sleeps done were 0, 0.4, 0.8, 1.6 and 3.2 sec -> total 6 sec. The new sleeping has no backoff factor and is the same length every time, so if we do 5 sleeps of 1 second each, then we get less overall delay, i.e. less likely to wait long enough for the server to become responsive again. We should use the config `rest_api.retry_interval` to control the delay.

File: egcg_core/rest_communication.py [hunk 1]
```diff
@@ -139,11 +139,16 @@ def _req(self, method, url, quiet=False, **kwargs):
...
             kwargs['data'] = kwargs.pop('json')
 
-        self.lock.acquire()
         try:
-            r = self.session.request(method, url, **kwargs)
-        finally:
-            self.lock.release()
+            with self.lock:
+                r = self.session.request(method, url, **kwargs)
+        except Exception as e:
```
Comment (mwhamgenomics) [inline egcg_core/rest_communication.py:145]: Is this too broad? The urllib3 Retry catches TimeoutError, HTTPException, SocketError, ProtocolError, BaseSSLError, SSLError and CertificateError.

Comment (tcezard) [reply egcg_core/rest_communication.py:145]: Because we would eventually raise it, I think it would be fine

**Review by tcezard — APPROVED** [review]

File: tests/test_rest_communication.py [hunk 2]
```diff
@@ -4,21 +4,17 @@
...
 from egcg_core.util import check_if_nested
 from egcg_core.exceptions import RestCommunicationError
 
 
-def rest_url(endpoint):
-    return 'http://localhost:4999/api/0.1/' + endpoint + '/'
-
-
 test_endpoint = 'an_endpoint'
+test_url = 'http://localhost:4999/api/0.1/' + test_endpoint + '/'
```
Comment (lukbut) [inline tests/test_rest_communication.py:17]: Maybe use urlparse?

Comment (mwhamgenomics) [reply tests/test_rest_communication.py:17]: urlparse is for splitting up a url into components, not building one. I could use urljoin and make it `urllib.parse.urljoin('http://localhost:4999', 'api/0.1/' + test_endpoint + '/')`, but it's not really worth it and I find the concatenated string more readable anyway

Comment (lukbut) [reply tests/test_rest_communication.py:17]: Noted.

**Review by lukbut — APPROVED** [review]

**Pull request merged**